crates/libm-test/benches/random.rs

```diff
@@ -0,0 +1,114 @@
+use std::hint::black_box;
+use std::time::Duration;
+
+use criterion::{Criterion, criterion_main};
+use libm_test::TupleCall;
+use libm_test::gen::random;
+
+/// Benchmark with this many items to get a variety
+const BENCH_ITER_ITEMS: usize = if cfg!(feature = "short-benchmarks") { 50 } else { 500 };
+
+macro_rules! musl_rand_benches {
+    (
+        fn_name: $fn_name:ident,
+        CFn: $CFn:ty,
+        CArgs: $CArgs:ty,
+        CRet: $CRet:ty,
+        RustFn: $RustFn:ty,
+        RustArgs: $RustArgs:ty,
+        RustRet: $RustRet:ty,
+        fn_extra: $skip_on_i586:expr,
+    ) => {
+        paste::paste! {
+            fn [< musl_bench_ $fn_name >](c: &mut Criterion) {
+                let fname = stringify!($fn_name);
+                let benchvec: Vec<_> = random::get_test_cases::<$RustArgs>(fname)
+                    .take(BENCH_ITER_ITEMS)
+                    .collect();
+
+                // Perform a sanity check that we are benchmarking the same thing
+                // Don't test against musl if it is not available
+                #[cfg(feature = "build-musl")]
+                for input in benchvec.iter().copied() {
+                    use anyhow::Context;
+                    use libm_test::{CheckBasis, CheckCtx, CheckOutput};
+
+                    const QUICKCHECK_ULP: u32 = 10;
+
+                    if cfg!(x86_no_sse) && $skip_on_i586 {
+                        break;
+                    }
+
+                    let musl_res = input.call(musl_math_sys::$fn_name as $CFn);
+                    let crate_res = input.call(libm::$fn_name as $RustFn);
+
+                    let ctx = CheckCtx::new(QUICKCHECK_ULP, fname, CheckBasis::Musl);
+                    musl_res.validate(crate_res, input, &ctx).context(fname).unwrap();
+                }
+
+                let mut group = c.benchmark_group(fname);
+                group.bench_function("crate", |b| b.iter(|| {
+                    for input in benchvec.iter().copied() {
+                        black_box(black_box(input).call(libm::$fn_name as $RustFn));
+                    }
```

Comment:
```suggestion
                    let f = black_box(libm::$fn_name as $RustFn);
                    for input in benchvec.iter().copied() {
                        input.call(f);
                    }
```
Forcing the call to be via function pointer would make the comparison fair. It's also much nicer for examining the benchmark with `perf` when the thing being measured hasn't been inlined into the benchmark-loop. With the called function becoming opaque, the inputs and outputs don't need `black_box`, which recovers some of the overhead.

Same changes for musl below.

Reply:
Thanks, applied to both.